ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/discovery/internal/RecursivePeerRefreshState.java

```diff
@@ -0,0 +1,167 @@
+import tech.pegasys.pantheon.util.bytes.BytesValue;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+class RecursivePeerRefreshState {
+  private final int CONCURRENT_REQUEST_LIMIT = 3;
+  private final BytesValue target;
+  private final PeerBlacklist peerBlacklist;
+  private final BondingAgent bondingAgent;
+  private final NeighborFinder neighborFinder;
+  private final HashMap<Peer, Integer> anteMap;
+  private final SortedSet<Map.Entry<Peer, Integer>> distanceSortedPeers;
+  private final List<OutstandingRequest> outstandingRequestList;
+  private final List<BytesValue> contactedInCurrentExecution;
+
+  RecursivePeerRefreshState(
+      final BytesValue target,
+      final PeerBlacklist peerBlacklist,
+      final BondingAgent bondingAgent,
+      final NeighborFinder neighborFinder) {
+    this.target = target;
+    this.peerBlacklist = peerBlacklist;
+    this.bondingAgent = bondingAgent;
+    this.neighborFinder = neighborFinder;
+    this.anteMap = new HashMap<>();
+    this.distanceSortedPeers = new TreeSet<>(Comparator.comparing(Map.Entry::getValue));
+    this.outstandingRequestList = new ArrayList<>();
+    this.contactedInCurrentExecution = new ArrayList<>();
+  }
+
+  /**
+   * The lookup initiator starts by picking α closest nodes to the target it knows of. The initiator
+   * then sends concurrent FindNode packets to those nodes. α is a system-wide concurrency
+   * parameter, such as 3.
+   */
+  private void initiatePeerRefreshCycle(final List<Peer> peers) {
+    for (Peer peer : peers) {
+      if (!contactedInCurrentExecution.contains(peer.getId())) {
+        BytesValue peerId = peer.getId();
+        outstandingRequestList.add(new OutstandingRequest(peerId));
+        contactedInCurrentExecution.add(peerId);
+        neighborFinder.issueFindNodeRequest(peer);
+      }
+      // The lookup terminates when the initiator has queried
+      // and gotten responses from the k closest nodes it has seen.
+    }
+  }
+
+  void digestNeighboursPacket(
+      final NeighborsPacketData neighboursPacket, final BytesValue peerIdentifier) {
+    if (outstandingRequestList.contains(new OutstandingRequest(peerIdentifier))) {
+      List<Peer> receivedPeerList = neighboursPacket.getNodes();
+      for (Peer receivedPeer : receivedPeerList) {
+        if (!peerBlacklist.contains(receivedPeer)) {
+          bondingAgent.performBonding(receivedPeer);
+          anteMap.put(receivedPeer, distance(target, receivedPeer.getId()));
+        }
+      }
+      outstandingRequestList.remove(new OutstandingRequest(peerIdentifier));
+      performIteration();
+    }
+  }
+
+  private List<Peer> determineFindNodeCandidates() {
+    distanceSortedPeers.addAll(anteMap.entrySet());
```

Comment:
Does this need to clear `distanceSortedPeers` first?  If we just want to have each peer in `anteMap` in a sorted list, it would be better to just iterate through `anteMap` to create an `ArrayList` then use `Collections.sort` to sort it in place rather than going via a sorted set.

Or, third option since we want a fixed number of items, iterate through `anteMap` adding each item to a `TreeMap` (need to create a custom class as we shouldn't be reusing `Map.Entry` instances`).  After adding each item check if the size of our set is > `CONCURRENT_REQUEST_LIMIT` if so use `pollLast` to remove the last item.  `pollLast` is from `NavigableMap`, a sub-type of `SortedSet` which `TreeSet` implements.